presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SimplifyExpressions.java

```diff
@@ -285,28 +277,21 @@ public Expression rewriteLogicalBinaryExpression(LogicalBinaryExpression node, N
         }
     }
 
-    private static class GroupingOperationContext
+    private static class GroupingOperationToConstant
+            extends ExpressionRewriter<Void>
     {
-        private final boolean isGroupIdNodePresent;
+        private final PlanNode source;
 
-        GroupingOperationContext(boolean isGroupIdNodePresent)
+        GroupingOperationToConstant(PlanNode source)
         {
-            this.isGroupIdNodePresent = isGroupIdNodePresent;
+            this.source = source;
         }
 
-        public boolean getIsGroupIdNodePresent()
-        {
-            return isGroupIdNodePresent;
-        }
-    }
-
-    private static class GroupingOperationToConstant
-            extends ExpressionRewriter<GroupingOperationContext>
-    {
         @Override
-        public Expression rewriteFunctionCall(FunctionCall node, GroupingOperationContext context, ExpressionTreeRewriter<GroupingOperationContext> treeRewriter)
+        public Expression rewriteFunctionCall(FunctionCall node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
         {
-            if (!context.getIsGroupIdNodePresent() && node.getName().toString().equals("grouping")) {
+            // A grouping() is preceded by a GroupId node when the source node is projecting a groupid symbol
+            if (!source.getOutputSymbols().contains(new Symbol("groupid")) && node.getName().toString().equals(GROUPING)) {
```

Comment:
I refactored the symbol stuff quite a bit. There can only be one `GroupIdNode` per `QuerySpecification` so I expanded `Analysis` to keep track of the `groupId` symbol for every `QuerySpecification`. This makes the rewrite in `GroupingOperationRewriter` ironclad. However, during the rewrites in `SimplifyExpressions` and `SymbolToInputRewriter` I couldn't think of an easy solution to extract the symbol so what I end up doing is substring comparison. I know this is still not ideal but before I embark on a more complicated solution I'd like to get Martin's thoughts to make sure I don't spend time building something that's wrong. Let me know what you think.